### DISCUSSION TITLE
BPMP/bootloader changes

### DISCUSSION THREAD

**Discussion by goetzdd** [OE4T/tegra-demo-distro#196] [General]
So In desiring to use the onboard serial ports I have issues that appear to be related to the BPMP stealing one of them.  As per this:

https://forums.developer.nvidia.com/t/tx2-nx-rs232-communication-problem/173226/10

The solution is to change the device tree for the bootloader but I suspect OE4T is not building the bootloader and I may have to use the stock l4t to produce bpmp.bin? If that's even the right thing to be going for.

**Comment by goetzdd** [comment]
Hmm...the reference nvidia link may be referring to just changing the dtb, maybe a bbappend that replaces the dtb is all I need? But to what recipe is the question.

**Comment by goetzdd** [comment]
Ok, verified just editing the bpmp dtb in the tegraflash files does fix this issue, but any pointers on how to structure this so as to be repeatable within the yocto/meta-tegra framework? Having to manually decompile/edit/recompile the dtb is time consuming and error prone.

**Reply by madisongh** [reply]
One way would be to create a custom machine configuration; you could store the modified DTB file somewhere (under a different name) and add bbappends for the `tegra-bootfiles` and `tegra-flashvars` recipes to use the modified copy instead of the standard one.  There's a [wiki page](https://github.com/OE4T/meta-tegra/wiki/Creating-a-custom-MACHINE) that has some basics on custom machines, but there is no step-by-step walkthrough.

For just modifying the existing DTB on the fly, you could do that in either a `tegra-bootfiles` bbappend (with a `do_install:append()` function) or by creating a `tegraflash_custom_pre()` function in your image recipe.  You could use `fdtput` to make the modification:
```
$ fdtput -r <name-of-dtb-file> /serial
```
The `tegraflash` image generation step already has a dependency on `dtc-native` (that's the recipe that provides the `fdtput` tool); if you use the `tegra-bootfiles` bbappend approach, make sure you add that dependency there.

**Comment by dat-tran-matternet** [comment]
I had to do this as well. I modified tegra-binaries to get this to overwrite the existing dtb file:

tegra-binaries_%.bbappend:

```
FILESEXTRAPATHS_prepend := "${THISDIR}/files:"

SRC_URI_append_jetson-xavier-nx-devkit-tx2-nx = "\
	file://tegra186-bpmp-p3636-0001-a00-00.dtb;subdir=Linux_for_Tegra/bootloader/t186ref \
"
```

@goetzdd let me know if that works for you.
